### ISSUE TITLE
Variable naming

### ISSUE THREAD

**Issue by rhfogh** [mxcube/HardwareRepository#589]
Could we agree some general guidelines for variable and function names?

I think it is very useful (particularly for code you did not write yourself but need to understand) if you can see from variable names not just 'what it is about', but also the variable type. This is too complex to be rigid about it, but we can try to make it obvious in each case what we are doing. 

Basically we should be able to distinguish between things, collections of things, names of things, actions, and some special things like maps or Booleans.

- variables and parameters:
    - names of objects and values are *singular*
    - names of collections are *plural*
    - names of maps and dictionaries are plural, or contain 'map', 'dict','data',  or an internel '2', like 'name2state'
- functions names should generally contain a verb
 - Numbers of things can be distinguished by names like 'num_samples' or 'sample_count', or 'len_args'
- Distinguish between objects and strings that refer to them. The name of a motor should be 'motor_name', not 'motor'; 'state' should be a state object; use e.g. 'name' for its string name.

- Booleans are tricky. They can be distinguished by 'is_' or 'has_' or 'use_' prefixes, but this is not always needed (e.g. read_only' is unmistakable by itself. Also I can find no clear distinction between Boolean attributes and Boolean functions (like 'is_ready'). Do we want to prohibit Boolean functions and make e.g. is_ready into a property, or is this taking things too far?

**Comment by marcus-oscarsson** [comment]
Sounds good to me as long as it does not get to complicated to remember, so you don't have to think for 15 minutes before naming a variable. Most of your suggestions seems quite natural to me so I'm fine with them as a general guideline

**Comment by marcus-oscarsson** [comment]
Actually I think we have already been over this one or two time and that we have already said that something like the above would be the recommended way, or am I wrong ?

**Comment by marcus-oscarsson** [comment]
Actually I think we have already been over this one or two time and that we have already said that something like the above would be the recommended way, or am I wrong ?

**Comment by marcus-oscarsson** [comment]
Actually I think we have already been over this one or two time and that we have already said that something like the above would be the recommended way, or am I wrong ?

**Comment by marcus-oscarsson** [comment]
Actually I think we have already been over this one or two time and that we have already said that something like the above would be the recommended way, or am I wrong ?

**Comment by marcus-oscarsson** [comment]
Actually I think we have already been over this one or two time and that we have already said that something like the above would be the recommended way, or am I wrong ?

**Comment by marcus-oscarsson** [comment]
Actually I think we have already been over this one or two time and that we have already said that something like the above would be the recommended way, or am I wrong ?

**Comment by rhfogh** [comment]
@marcus-oscarsson 
I do not remember if we ever had any official agreement on this point, butall the better if we do.

Anyway, I came to this because of two examples:
  - using the variable 'state' for a list of state names
  - using the parameter 'args' to hold a single string that evaluates to an int.
The discussion in the PRs did not suggest that this usage was against any shared agreement, so I thought we maybe needed to (re)visit the point.

**Comment by marcus-oscarsson** [comment]
I see, I'm pretty sure the case is that most of us is trying to get things to work and sometimes it just ends up breaking some of these sound guidelines. Sometimes we just want to add a simple functionality without rewriting large bits of code

**Comment by beteva** [comment]
You are right @rhfogh, we could, indeed, add few rules to the style guide we've already have adopted and described in the contributing file. I agree with @marcus-oscarsson that this should not get too constraining. For me the most important is that the methods are documented and whenever possible the variable name mentions its type (like var_list, var_dict, var_hwobj).
As for the choice of property or function, it will be guided by the number of input arguments. Whenever we need more than one, it could only be a function.For boolean, I tend to prefer a property, but it is not a "strong opinion".

**Comment by beteva** [comment]
You are right @rhfogh, we could, indeed, add few rules to the style guide we've already have adopted and described in the contributing file. I agree with @marcus-oscarsson that this should not get too constraining. For me the most important is that the methods are documented and whenever possible the variable name mentions its type (like var_list, var_dict, var_hwobj).
As for the choice of property or function, it will be guided by the number of input arguments. Whenever we need more than one, it could only be a function.For boolean, I tend to prefer a property, but it is not a "strong opinion".

**Comment by beteva** [comment]
You are right @rhfogh, we could, indeed, add few rules to the style guide we've already have adopted and described in the contributing file. I agree with @marcus-oscarsson that this should not get too constraining. For me the most important is that the methods are documented and whenever possible the variable name mentions its type (like var_list, var_dict, var_hwobj).
As for the choice of property or function, it will be guided by the number of input arguments. Whenever we need more than one, it could only be a function.For boolean, I tend to prefer a property, but it is not a "strong opinion".

**Comment by marcus-oscarsson** [comment]
I've started to add a few of the ideas mentioned by @rhfogh in the coding [convention](https://github.com/mxcube/HardwareRepository/blob/master/CONTRIBUTING.md). We will discuss them in more detail tomorow.
